core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala

```diff
@@ -1173,147 +1175,265 @@ class UserServiceSpec
-          Set.empty,
-          Set.empty
-        )
-      )
-      val samDAO = mock[SamDAO](RETURNS_SMART_NULLS)
-      when(samDAO.listUserResources(SamResourceTypeNames.billingProject, userInfo))
-        .thenReturn(Future.successful(userBillingResources))
-      val bpmDAO = mock[BillingProfileManagerDAO](RETURNS_SMART_NULLS)
-      when(
-        bpmDAO.getHardcodedAzureBillingProject(ArgumentMatchers.eq(userBillingResources.map(_.resourceId).toSet),
-                                               ArgumentMatchers.eq(userInfo)
-        )(any())
+    val billingProjects = Set(ownerProject, billingProfileBackedProject, hardcodedExternalProject)
+
+    val repository = mock[BillingRepository]
+    when(repository.getBillingProjects(ArgumentMatchers.eq(billingProjects.map(_.projectName))))
+      .thenReturn(Future.successful(Seq(ownerProject, billingProfileBackedProject)))
+
+    // Setup mock DAOs
+    val ownerRole = SamRolesAndActions(Set(SamBillingProjectRoles.owner), Set(SamBillingProjectActions.createWorkspace))
+    val noRole = SamRolesAndActions(Set.empty, Set.empty)
+    val creatorRole =
+      SamRolesAndActions(Set(SamBillingProjectRoles.workspaceCreator), Set(SamBillingProjectActions.createWorkspace))
+    val userBillingResources = Seq(
+      SamUserResource(ownerProject.projectName.value, ownerRole, noRole, noRole, Set.empty, Set.empty),
+      SamUserResource(billingProfileBackedProject.projectName.value, creatorRole, noRole, noRole, Set.empty, Set.empty),
+      SamUserResource(hardcodedExternalProject.projectName.value, creatorRole, noRole, noRole, Set.empty, Set.empty)
+    )
+    val samDAO = mock[SamDAO](RETURNS_SMART_NULLS)
+    when(samDAO.listUserResources(SamResourceTypeNames.billingProject, userInfo))
+      .thenReturn(Future.successful(userBillingResources))
+    val bpmDAO = mock[BillingProfileManagerDAO](RETURNS_SMART_NULLS)
+    when(bpmDAO.getAllBillingProfiles(testContext)).thenReturn(Future.successful(Seq(bpmBillingProfile)))
+    when(
+      bpmDAO.getHardcodedAzureBillingProject(ArgumentMatchers.eq(userBillingResources.map(_.resourceId).toSet), any())(
+        any()
       )
-        .thenReturn(Future.successful(Seq.empty))
-      when(bpmDAO.getAllBillingProfiles(testContext)).thenReturn(Future.successful(Seq.empty))
+    ).thenReturn(Future.successful(Seq(hardcodedExternalProject)))
 
-      val userService = getUserService(dataSource, samDAO, billingProfileManagerDAO = bpmDAO)
+    val userService = getUserService(samDAO = samDAO, bpmDAO = bpmDAO, billingRepository = Some(repository))
 
-      val result = Await.result(userService.listBillingProjectsV2(), Duration.Inf)
+    val expected = Seq(
+      RawlsBillingProjectResponse(Set(ProjectRoles.Owner), ownerProject, CloudPlatform.GCP),
+      RawlsBillingProjectResponse(Set(ProjectRoles.User), billingProfileBackedProject, CloudPlatform.AZURE),
+      RawlsBillingProjectResponse(Set(ProjectRoles.User), hardcodedExternalProject, CloudPlatform.GCP)
```

Comment:
Nit: the hardcoded project really shouldn't be GCP. But it's going away soon, so it doesn't really matter.